### ISSUE TITLE
`job-info.eventlog-watch` on `guest.output` hangs for jobs before `start` event, then returns ENOENT after job starts

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#5344]
Reproducer:

 1. submit a held job: `flux submit --urgency=0 hostname`
 2. try to watch output eventlog: `flux job wait-event -p guest.output JOBID data` (hangs)
 3. release job: `flux job urgency JOBID default`
 4. `flux job wait-event` returns with error: `flux-job: eventlog path guest.output not found`

It would probably be best to return an error immediately or wait for the eventlog to appear or the job to become inactive, then return an error.

**Comment by chu11** [comment]
I believe this is a duplicate of #5301, that small window between guest namespace creation and shell creating `guest.output` leads to the ENOENT.  It's a good reproducer case to add a test.

**Comment by grondo** [comment]
Until #5301 is fixed, this is a separate bug. The RPC should return ENOENT immediately and not hang until the job starts.

In fact, if there was some other eventlog being watched under the guest namespace, would it hit this same hang? Does `job-info` block waiting for any path under the guest namespace currently?

**Comment by chu11** [comment]
> In fact, if there was some other eventlog being watched under the guest namespace, would it hit this same hang? Does job-info block waiting for any path under the guest namespace currently?

I believe so.  For any path under "guest." it waits until the namespace exists.  So the initial hang makes sense at the moment.  I guess the thought process when `job-info` was written was that a request for "guest." inherently wants to wait for the guest namespace to exist.

> The RPC should return ENOENT immediately and not hang until the job starts.

I don't think this is possible right now, as waiting for the guest namespace to exist is part of the original design.

Ahhh, re-reading above you stated "return an error immediately or wait for the eventlog to appear or the job to become inactive, ***then*** return an error."

Several corner cases I'm now thinking about:

```
jobid=`flux submit --urgency=hold sleep inf`
flux job wait-event -p guest.output $jobid foobar &
flux cancel $jobid
```

in this case, the guest namespace can never happen.  So the wait-event presumably should fail or minimally exit at some point.  I'm not sure if this is handled or not right now.

More generically I'm thinking of something like this (after #5301 is fixed):

```
jobid=`flux submit --urgency=hold sleep inf`
flux job wait-event -p guest.output $jobid foobar &
flux job urgency $jobid default
```

in this impossible case (event "foobar"), right now I think behavior is to hang, i.e. never return on an event that can never exist.  Thus the timeout option that exists with `flux job wait-event`.

But for what you mentioned above

```
jobid=`flux submit --urgency=hold sleep inf`
flux job wait-event -p guest.INVALIDPATH $jobid foobar &
flux job urgency $jobid default
```

you are correct, we should probably return ENOENT when the job becomes inactive.

Gotta ponder on this ... will re-open

**Comment by chu11** [comment]
>  Yeah, I'm wondering if this was the right decision. At this point, anyway, it feels like having job-info return ENOENT for any key that doesn't exist at the time of the call (even if under the guest namespace) would be a lot more consistent and have less corner cases (like those listed above and more) and thus be much easier to reason about.

I think the primary use case was going to be something like `flux job wait-event -p guest.exec.eventlog FOO` on a pending job.  Which I think waiting for the guest namespace to exist would be a reasonable thing to expect.  I see in the testsuite we do things like:

```
jobid=`flux job submit SOMETHING` &&
flux job wait-event -p guest.exec.eventlog SOMETHING
```

which I believe will inherently become racy if we remove the waiting for guest namespace?

> I'm also wondering if even waiting for the output eventlog makes sense. I think everything in flux-core anyway already waits for the shell.init event before trying to read output, so maybe automatically waiting for the output eventlog is actually a misstep?

I think you're referring to #5301 here?  As I was working on #5301 I began to agree with this, thinking a `--waitcreate` would be better than automatically doing it.

> For example, flux job info JOBID R will currently return an error immediately for a pending job. To be consistent, should it not block until R exists like it does for all guest namespace entries? To be consistent should it block on any job key, only returning ENOENT if the job becomes inactive and the key still does not exist?

To me I think there is a difference between `flux job info` and `flux job wait-event`, where the former is "lookup now" and the latter "wait for an event", so waiting for an eventlog to exist feels like a reasonable thing to expect from the "wait".

BUT perhaps it need not be the default?  OR there should be an option to not do that?  OR perhaps the wait only exists by default for special cases such as documented eventlogs? i.e. anything in guest does not get special treatment?

**Comment by chu11** [comment]
Been working on #5301 and it ends up that many of the corner cases I mention about are already handled in `job-info`.  The "waitcreate" is the only issue.  So I'm going to re-close this.
